### DISCUSSION TITLE
Running Spinetoolbox from command line.

### DISCUSSION THREAD

**Discussion by IngridJSJ** [Spine-project/Spine-Toolbox#1736] [Discussion]
Hello, 
I have now access to a HPC and would like to know if I can run my project there. 
So, is it possible to run spinetoolbox worfklow as a script?

Also I tried to install spinetoolbox in linux but it didnt run as smoothly. 
when trying to run 
 python -m pipx install git+https://github.com/Spine-project/spinetoolbox-dev
I got these errors

Building wheel for pyodbc (setup.py) ... error
  ERROR: Command errored out with exit status 1:
   command: /home/isanchezjimene/.conda/envs/spinetoolbox-dev/bin/python -u -c 'import io, os, sys, setuptools, tokenize; sys.argv[0] = '"'"'/tmp/pip-install-l536ksqz/pyodbc_83bd7cf51b4e4bc197a59d7c90436bd9/setup.py'"'"'; __file__='"'"'/tmp/pip-install-l536ksqz/pyodbc_83bd7cf51b4e4bc197a59d7c90436bd9/setup.py'"'"';f = getattr(tokenize, '"'"'open'"'"', open)(__file__) if os.path.exists(__file__) else io.StringIO('"'"'from setuptools import setup; setup()'"'"');code = f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' bdist_wheel -d /tmp/pip-wheel-4u7x4v36
       cwd: /tmp/pip-install-l536ksqz/pyodbc_83bd7cf51b4e4bc197a59d7c90436bd9/
  Complete output (14 lines):
  running bdist_wheel
  running build
  running build_ext
  creating build
  creating build/temp.linux-x86_64-3.8
  creating build/temp.linux-x86_64-3.8/src
  /apps/arch/2022r2/software/linux-rhel8-skylake_avx512/gcc-8.5.0/gcc-11.2.0-midhpa4emjkjgec276wnpdtxr7f4i5rb/bin/gcc -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -fPIC -DPYODBC_VERSION=4.0.32 -I/home/isanchezjimene/.conda/envs/spinetoolbox-dev/include/python3.8 -c src/buffer.cpp -o build/temp.linux-x86_64-3.8/src/buffer.o -Wno-write-strings -DHAVE_UNISTD_H -DHAVE_PWD_H -DHAVE_SYS_TYPES_H -DHAVE_LONG_LONG -DSIZEOF_LONG_INT=8 -I/usr/include
  cc1plus: warning: command-line option '-Wstrict-prototypes' is valid for C/ObjC but not for C++
  In file included from src/buffer.cpp:12:
  src/pyodbc.h:56:10: fatal error: sql.h: No such file or directory
     56 | #include <sql.h>
        |          ^~~~~~~
  compilation terminated.
  error: command '/apps/arch/2022r2/software/linux-rhel8-skylake_avx512/gcc-8.5.0/gcc-11.2.0-midhpa4emjkjgec276wnpdtxr7f4i5rb/bin/gcc' failed with exit code 1
  ----------------------------------------
  ERROR: Failed building wheel for pyodbc
  Running setup.py clean for pyodbc
Failed to build pyodbc
Installing collected packages: spinedb-api, spine-engine, pyodbc, spinetoolbox, spine-items
  Attempting uninstall: spinedb-api
    Found existing installation: spinedb-api 0.22.0.dev0
    Uninstalling spinedb-api-0.22.0.dev0:
      Successfully uninstalled spinedb-api-0.22.0.dev0
  Running setup.py develop for spinedb-api
  Attempting uninstall: spine-engine
    Found existing installation: spine-engine 0.17.0.dev0
    Uninstalling spine-engine-0.17.0.dev0:
      Successfully uninstalled spine-engine-0.17.0.dev0
  Running setup.py develop for spine-engine
    Running setup.py install for pyodbc ... error
    ERROR: Command errored out with exit status 1:
     command: /home/isanchezjimene/.conda/envs/spinetoolbox-dev/bin/python -u -c 'import io, os, sys, setuptools, tokenize; sys.argv[0] = '"'"'/tmp/pip-install-l536ksqz/pyodbc_83bd7cf51b4e4bc197a59d7c90436bd9/setup.py'"'"'; __file__='"'"'/tmp/pip-install-l536ksqz/pyodbc_83bd7cf51b4e4bc197a59d7c90436bd9/setup.py'"'"';f = getattr(tokenize, '"'"'open'"'"', open)(__file__) if os.path.exists(__file__) else io.StringIO('"'"'from setuptools import setup; setup()'"'"');code = f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' install --record /tmp/pip-record-8053f6n0/install-record.txt --single-version-externally-managed --compile --install-headers /home/isanchezjimene/.conda/envs/spinetoolbox-dev/include/python3.8/pyodbc
         cwd: /tmp/pip-install-l536ksqz/pyodbc_83bd7cf51b4e4bc197a59d7c90436bd9/
    Complete output (14 lines):
    running install
    running build
    running build_ext
    creating build
    creating build/temp.linux-x86_64-3.8
    creating build/temp.linux-x86_64-3.8/src
    /apps/arch/2022r2/software/linux-rhel8-skylake_avx512/gcc-8.5.0/gcc-11.2.0-midhpa4emjkjgec276wnpdtxr7f4i5rb/bin/gcc -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -fPIC -DPYODBC_VERSION=4.0.32 -I/home/isanchezjimene/.conda/envs/spinetoolbox-dev/include/python3.8 -c src/buffer.cpp -o build/temp.linux-x86_64-3.8/src/buffer.o -Wno-write-strings -DHAVE_UNISTD_H -DHAVE_PWD_H -DHAVE_SYS_TYPES_H -DHAVE_LONG_LONG -DSIZEOF_LONG_INT=8 -I/usr/include
    cc1plus: warning: command-line option '-Wstrict-prototypes' is valid for C/ObjC but not for C++
    In file included from src/buffer.cpp:12:
    src/pyodbc.h:56:10: fatal error: sql.h: No such file or directory
       56 | #include <sql.h>
          |          ^~~~~~~
    compilation terminated.
    error: command '/apps/arch/2022r2/software/linux-rhel8-skylake_avx512/gcc-8.5.0/gcc-11.2.0-midhpa4emjkjgec276wnpdtxr7f4i5rb/bin/gcc' failed with exit code 1
    ----------------------------------------
ERROR: Command errored out with exit status 1: /home/isanchezjimene/.conda/envs/spinetoolbox-dev/bin/python -u -c 'import io, os, sys, setuptools, tokenize; sys.argv[0] = '"'"'/tmp/pip-install-l536ksqz/pyodbc_83bd7cf51b4e4bc197a59d7c90436bd9/setup.py'"'"'; __file__='"'"'/tmp/pip-install-l536ksqz/pyodbc_83bd7cf51b4e4bc197a59d7c90436bd9/setup.py'"'"';f = getattr(tokenize, '"'"'open'"'"', open)(__file__) if os.path.exists(__file__) else io.StringIO('"'"'from setuptools import setup; setup()'"'"');code = f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' install --record /tmp/pip-record-8053f6n0/install-record.txt --single-version-externally-managed --compile --install-headers /home/isanchezjimene/.conda/envs/spinetoolbox-dev/include/python3.8/pyodbc Check the logs for full command output.

So is there a way to run smoothly spinetoolbox in linux?

**Comment by soininen** [comment]
Hi Ingrid,

If you've installed Toolbox on the remote server, and the project directory is available there too, you can execute the project from the command line in "headless" mode without opening the GUI:

```
python -mspinetoolbox --execute-only /path/to/project-dir/
```

For more information on the command line options such as how to select specific project items for execution, run

```
python -mspinetoolbox --help
```

**Comment by soininen** [comment]
The installation problem is not really a Toolbox issue but it looks like your system is missing some packages. You could try searching the web for `src/pyodbc.h:56:10: fatal error: sql.h: No such file or directory` or maybe [this](https://github.com/mkleehammer/pyodbc/issues/441) discussion provides some help.

**Comment by ptsavol** [comment]
There's this in the **Troubleshooting** section of Spine Toolbox's readme.md.

```
Installation fails on Linux

If Python runs into errors while installing on Linux systems, running the following commands in a terminal may help:

$ sudo apt install libpq-dev
$ sudo apt-get install unixodbc-dev
```

**Reply by IngridJSJ** [reply]
Thank you for the answers. 
I tried the sudo apt-get install unixodbc-dev but unfortunately this command is not allowed in the HPC... 
I will try other things and let you know
